idea/idea-fir/src/org/jetbrains/kotlin/idea/quickfix/fixes/ReplaceCallFixFactories.kt

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.idea.quickfix.fixes
+
+import com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.idea.frontend.api.symbols.markers.KtTypedSymbol
+import org.jetbrains.kotlin.idea.frontend.api.types.KtTypeNullability
+import org.jetbrains.kotlin.idea.frontend.api.types.KtTypeWithNullability
+import org.jetbrains.kotlin.idea.quickfix.ReplaceImplicitReceiverCallFix
+import org.jetbrains.kotlin.idea.quickfix.ReplaceWithSafeCallFix
+import org.jetbrains.kotlin.idea.quickfix.quickFixesPsiBasedFactoryWithAnalysis
+import org.jetbrains.kotlin.idea.references.mainReference
+import org.jetbrains.kotlin.psi.*
+
+object ReplaceCallFixFactories {
+    val UnsafeCallFactory = quickFixesPsiBasedFactoryWithAnalysis { psiElement: PsiElement ->
+        fun KtExpression.shouldHaveNotNullType(): Boolean {
+            // This function is used to determine if we may need to add an elvis operator after the safe call. For example, to replace
+            // `s.length` in `1 + s.length` with a safe call, it should be replaced with `(s.length ?: <caret>)`.
+            // NOTE: Even though we could, we don't check to see if the call is an argument and if the parameter type is nullable.
+            // FE1.0 doesn't either (see ReplaceWithSafeCallFixFactory).
+            val type = when (val parent = parent) {
+                is KtBinaryExpression -> parent.left?.getKtType()
+                is KtProperty -> {
+                    // Case: `val i = s.length`. `parent.getReturnKtType()` would return Int and therefore an elvis would be added,
+                    // which does not match FE1.0 behavior. We want to see if there is an explicit type, e.g., `val i: Int = s.length`
+                    // and default to the replacement being nullable if there is no explicit type.
+                    if (parent.typeReference != null) {
+                        parent.getReturnKtType()
+                    } else null
+                }
+                else -> null
+            }
+            return (type as? KtTypeWithNullability)?.nullability == KtTypeNullability.NON_NULLABLE
+        }
+
+        if (psiElement is KtDotQualifiedExpression) {
+            listOf(ReplaceWithSafeCallFix(psiElement, psiElement.shouldHaveNotNullType()))
+        } else {
+            if (psiElement is KtNameReferenceExpression) {
+                // TODO: Determine if the call has an implicit receiver -- is this needed? See ReplaceWithSafeCallFixFactory in FE1.0.
```

Comment:
It seems the check is indeed unnecessary since for a explicitly qualified call, FIR is always reporitng the error on the `KtDotQualifiedExpression`. Can you mention this in this TODO? Something like "confirm the check is indeed not needed because ..."

Reply:
The check `is KtNameReferenceExpression` is necessary because we don't want to accidentally invoke the fix on an expression that is neither dot-qualified nor implicit receiver.

What the comment is referring to is that in FE1.0, there's an additional check `psiElement.getResolvedCall(psiElement.analyze())?.getImplicitReceiverValue() != null`. Right now there's no way to do this in the high-level API. But at the same time, I don't _think_ it is needed with the current code, hence the "-- is this needed?" in my comment. I don't think it is strictly necessary, I think it is just a safety precaution in case `UNSAFE_CALL` is accidentally reported on some random `KtNameReferenceExpression` that is not a call.

Reply:
I updated the comment to make it clearer. (hopefully!)

Reply:
Yes, I was talking about that check you left out here but present in FE1.0. I believe it's indeed unnecessary because for a explicitly qualified call, FIR is always reporitng the error on the KtDotQualifiedExpression. Can you update the TODO to mention it here so it's clear that this check was deliberately left out?

Reply:
Not sure why we're discussing explicitly qualified call here, since that's in the previous branch (`is KtDotQualifiedExpression`)... AFAICT the check for call receiver in FE1.0 is necessary in case `UNSAFE_CALL` is reported on some `KtNameReferenceExpression` that is not a call. E.g., you just put some random unresolved identifier `asdfghjkl` in the code which is a `KtNameReferenceExpression` but the checker incorrectly reports `UNSAFE_CALL` on it.

So I didn't leave the check for implicit receiver out because it's useless. It serves a purpose in FE1.0, although the purpose very limited. I would have added it here too for FIR, but it just was not possible currently with the high-level API.

For what it's worth, if I removed the check in FE1.0, all the quick fix tests pass. Which does make me think it's if limited value or uesless, but since test data by nature is limited in coverage, I don't want to take it out.

Reply:
IIUC FIR would only report UNSAFE_CALL on a `KtNameReferenceExpression`. this is guaranteed by https://github.com/JetBrains/kotlin/commit/0accaf0f30110f4eb963a9e59a72805ff9ab8264#diff-9177f36c606c764bbcf6b938b75cb5466eb0b221492983a216543bc9a0fcd0aeR37. Therefore, I think you can mention this point in this TODO here.

TBH, I think your current implementation is correct and this TODO is unnecessary. A note explaining why the check is not needed should be sufficient.